File: apps/plugins/jetbrains/src/test/kotlin/ai/yet/plugin/bridge/ControlledFileReadTest.kt
package ai.yet.plugin.bridge

import ai.yet.plugin.identity.ProductIdentity
import ai.yet.plugin.ui.JetBrainsControlledFileReadBridge
import com.google.gson.JsonParser
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ControlledFileReadTest {
    @Test
    fun parsesExplicitSafeRequest() {
        val request = assertNotNull(ControlledFileRead.parse(validRequest()))

        assertEquals("gui-s83-read-request", request.requestId)
        assertEquals("workspace-s83-request", request.controlledWorkspaceId)
        assertEquals("docs/architecture/013-agent-readiness-milestone.md", request.workspaceRelativePath)
        assertEquals(2048, request.maxBytes)
        assertEquals(80, request.maxLines)
        assertEquals(true, request.allowBody)
    }

    @Test
    fun rejectsUnsafeReadRequests() {
        assertNull(ControlledFileRead.parse(validRequest(path = "../README.md")))
        assertNull(ControlledFileRead.parse(validRequest(path = "node_modules/pkg/index.js")))
        assertNull(ControlledFileRead.parse(validRequest(path = "docs/.hidden.md")))
        assertNull(ControlledFileRead.parse(validRequest(requestId = "provider_key")))
        assertNull(ControlledFileRead.parse(validRequest(extra = ""","command":"cat""")))
        assertNull(ControlledFileRead.parse(validRequest(maxBytes = 8193)))
        assertNull(ControlledFileRead.parse(validRequest(recursive = true)))
    }

    @Test
    fun unsupportedResultIsFailClosedAndSanitized() {
        val request = assertNotNull(ControlledFileRead.parse(validRequest()))
        val message = JsonParser.parseString(ControlledFileRead.unsupportedResult(request)).asJsonObject
        val payload = message.getAsJsonObject("payload")
        val workspace = payload.getAsJsonObject("workspace")
        val readRequest = payload.getAsJsonObject("request")
        val budget = readRequest.getAsJsonObject("budget")
        val policy = payload.getAsJsonObject("policyFlags")
        val result = payload.getAsJsonObject("result")

        assertEquals(ProductIdentity.bridgeVersion, message.get("version").asString)
        assertEquals("host.controlledAgentFileReadResult", message.get("type").asString)
        assertEquals("gui-s83-read-request", message.get("requestId").asString)
        assertEquals("controlled_agent_file_read", payload.get("kind").asString)
        assertEquals("jetbrains", workspace.get("host").asString)
        assertEquals(false, workspace.get("privatePathExposed").asBoolean)
        assertEquals("docs/architecture/013-agent-readiness-milestone.md", readRequest.get("workspaceRelativePath").asString)
        assertEquals(false, budget.get("allowBody").asBoolean)
        assertEquals(false, policy.get("fileReadAllowed").asBoolean)
        assertEquals(false, policy.get("shellAllowed").asBoolean)
        assertEquals(false, policy.get("gitAllowed").asBoolean)
        assertEquals(false, policy.get("providerAllowed").asBoolean)
        assertEquals("disabled", result.get("status").asString)
        assertEquals("read_disabled", result.get("blockedReason").asString)
        assertEquals(false, result.get("bodyIncluded").asBoolean)
        assertFalse(result.has("text"))
        assertFalse(ControlledFileRead.unsupportedResult(request).contains("/Users/"))
    }

    @Test
    fun bridgeHandlesOnlyExplicitControlledReadAndReturnsUnsupported() {
        val sent = mutableListOf<String>()
        val handled = JetBrainsControlledFileReadBridge.handleControlledFileReadRequest(validRequest(), sent::add)

        assertTrue(handled)
        assertEquals(1, sent.size)
        val result = JsonParser.parseString(sent.single()).asJsonObject.getAsJsonObject("payload").getAsJsonObject("result")
        assertEquals("disabled", result.get("status").asString)
        assertEquals(false, result.get("bodyIncluded").asBoolean)
        assertFalse(JetBrainsControlledFileReadBridge.handleControlledFileReadRequest("""{"version":"${ProductIdentity.bridgeVersion}","type":"gui.ready"}""", sent::add))
    }

    private fun validRequest(
        requestId: String = "gui-s83-read-request",
        path: String = "docs/architecture/013-agent-readiness-milestone.md",
        maxBytes: Int = 2048,
        recursive: Boolean = false,
        extra: String = "",
    ): String = """
        {
          "version": "${ProductIdentity.bridgeVersion}",
          "type": "gui.controlledAgentFileReadRequest",
          "requestId": "$requestId",
          "payload": {
            "requestIdMintedBy": "gui",
            "source": "gui",
            "assistantMinted": false,
            "controlledWorkspaceId": "workspace-s83-request",
            "runId": "run-s83-request",
            "runtimeSessionId": "runtime-s83-request",
            "sessionId": "session-s83-request",
            "workspaceRelativePath": "$path",
            "maxBytes": $maxBytes,
            "maxLines": 80,
            "allowBody": true,
            "singleFileOnly": true,
            "recursive": $recursive,
            "globAllowed": false,
            "regexAllowed": false,
            "indexingAllowed": false
            $extra
          }
        }
    """.trimIndent()
}

File: apps/plugins/vscode/src/controlledFileRead.test.ts
import * as assert from "node:assert/strict";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { runControlledFileReadRequest } from "./controlledFileRead";

async function main(): Promise<void> {
  await testSafeRead();
  await testUnsafePathsFailClosed();
  await testBinaryOversizeSymlinkAndSecretBodiesFailClosed();
}

async function testSafeRead(): Promise<void> {
  const workspace = await createWorkspace();
  await fs.mkdir(path.join(workspace, "docs"), { recursive: true });
  await fs.writeFile(path.join(workspace, "docs", "note.md"), "hello\nthere\n", "utf8");

  const result = await runControlledFileReadRequest(createRequest("docs/note.md"), [workspace]);

  assert.equal(result.type, "host.controlledAgentFileReadResult");
  assert.equal(result.requestId, "req-read-safe");
  assert.equal(result.payload.result.status, "success");
  assert.equal(result.payload.result.text, "hello\nthere\n");
  assert.equal(result.payload.result.sanitizedPathLabel, "docs/note.md");
  assert.equal(result.payload.result.bodyIncluded, true);
  assert.equal(result.payload.workspace.privatePathExposed, false);
  assert.equal(JSON.stringify(result).includes(workspace), false);
}

async function testUnsafePathsFailClosed(): Promise<void> {
  const workspace = await createWorkspace();
  await fs.mkdir(path.join(workspace, "src"), { recursive: true });
  await fs.writeFile(path.join(workspace, "src", "main.ts"), "export const ok = true;\n", "utf8");

  for (const unsafePath of ["/src/main.ts", "../src/main.ts", "src\\main.ts", "src/.hidden.ts", "node_modules/pkg/index.js", "dist/app.js", "src/api_key.txt"]) {
    const result = await runControlledFileReadRequest(createRequest(unsafePath), [workspace]);
    assert.equal(result.payload.result.status, "blocked", unsafePath);
    assert.equal(result.payload.result.bodyIncluded, false, unsafePath);
    assert.equal("text" in result.payload.result, false, unsafePath);
  }
}

async function testBinaryOversizeSymlinkAndSecretBodiesFailClosed(): Promise<void> {
  const workspace = await createWorkspace();
  await fs.mkdir(path.join(workspace, "src"), { recursive: true });
  await fs.writeFile(path.join(workspace, "src", "bin.txt"), Buffer.from([0, 1, 2]));
  await fs.writeFile(path.join(workspace, "src", "large.txt"), "a".repeat(8193), "utf8");
  await fs.writeFile(path.join(workspace, "src", "marker.txt"), "password = nope\n", "utf8");
  await fs.symlink(path.join(workspace, "src", "marker.txt"), path.join(workspace, "src", "link.txt"));

  for (const [workspaceRelativePath, reason] of [["src/bin.txt", "binary_file"], ["src/large.txt", "too_large"], ["src/marker.txt", "policy_denied"], ["src/link.txt", "symlink_denied"]] as const) {
    const result = await runControlledFileReadRequest(createRequest(workspaceRelativePath), [workspace]);
    assert.equal(result.payload.result.status, "blocked", workspaceRelativePath);
    assert.equal(result.payload.result.blockedReason, reason, workspaceRelativePath);
    assert.equal(result.payload.result.bodyIncluded, false, workspaceRelativePath);
    assert.equal("text" in result.payload.result, false, workspaceRelativePath);
  }
}

async function createWorkspace(): Promise<string> {
  return fs.mkdtemp(path.join(os.tmpdir(), "yet-ai-controlled-read-"));
}

function createRequest(workspaceRelativePath: string): Parameters<typeof runControlledFileReadRequest>[0] {
  return {
    version: "2026-05-15",
    type: "gui.controlledAgentFileReadRequest",
    requestId: "req-read-safe",
    payload: {
      requestIdMintedBy: "gui",
      source: "gui",
      assistantMinted: false,
      controlledWorkspaceId: "workspace-read-safe",
      runId: "run-read-safe",
      runtimeSessionId: "runtime-read-safe",
      sessionId: "session-read-safe",
      workspaceRelativePath,
      maxBytes: 8192,
      maxLines: 240,
      allowBody: true,
      singleFileOnly: true,
      recursive: false,
      globAllowed: false,
      regexAllowed: false,
      indexingAllowed: false,
    },
  };
}

void main();
